lib/services/local/generic/nonce.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"context"
+	"errors"
+	"math"
+	"time"
+
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+// nonceProtectedResourceShim is a helper for quickly extracting the nonce
+type nonceProtectedResourceShim struct {
+	Spec struct {
+		Nonce uint64 `json:"nonce"`
+	} `json:"spec"`
+}
+
+// NonceViolation is the error returned by FastUpdateNonceProtectedResource when a nonce-protected
+// update fails due to concurrent modification. This error should be caught and re-mapped into an
+// appropriate user-facing message for the given resource type.
+var NonceViolation = errors.New("nonce-violation")
+
+// NonceProtectedResource describes the expected methods for a resource that is protected
+// from concurrent modification by a nonce.
+type NonceProtectedResource interface {
```

Comment:
Can we make this interfaces unexported. It looks like it is used only inside this package.